pkg/schedulers/batch_table_scheduler/batch_table_scheduler.go

```diff
@@ -324,90 +333,136 @@ func (scheduler *batchScheduler) Close() {
+		var round uint
+		var closing bool
+		latch := make(map[uint]uint)
+
+		flushFunc := func() {
+			var curBatch []*core.Msg
+			if len(batch) > scheduler.cfg.MaxBatchPerWorker {
+				curBatch = make([]*core.Msg, scheduler.cfg.MaxBatchPerWorker)
+				copy(curBatch, batch[:scheduler.cfg.MaxBatchPerWorker])
+			} else {
+				curBatch = make([]*core.Msg, len(batch))
+				copy(curBatch, batch)
+			}
+			for _, item := range curBatch {
+				hash := item.OutputHash()
+				if item.OutputStreamKey != core.NoDependencyOutput {
+					if latch[hash] > 0 {
+						return
 					}
-					// log.Printf("workingSet batch keys: %v\n", workingSetBatchKeys)
-					scheduler.workingSet.checkAndPutBatch(workingSetBatchKeys)
 				}
-				var realBatch []*core.Msg
-				for _, msgInBatch := range batch {
-					if msgInBatch.Type == core.MsgCtl {
-						err := scheduler.AckMsg(msgInBatch)
-						if err != nil {
-							log.Fatalf("[batch_scheduler] failed to ack slide window, SequenceNumber: %v, err: %v",
-								*msgInBatch.InputSequence, errors.ErrorStack(err))
-						}
-					} else {
-						realBatch = append(realBatch, msgInBatch)
+			}
+			for _, m := range curBatch {
+				h := m.OutputHash()
+				latch[h]++
+				oldCB := m.AfterAckCallback
+				m.AfterAckCallback = func() error {
+					latchC <- h
+					if oldCB != nil {
+						return oldCB()
 					}
+					return nil
+				}
+			}
+			queueIdx := round % uint(scheduler.cfg.NrWorker)
+			round++
+			if !closing {
+				scheduler.workerQueues[queueIdx] <- curBatch
+				if len(batch) > scheduler.cfg.MaxBatchPerWorker {
+					batch = batch[scheduler.cfg.MaxBatchPerWorker:]
+				} else if cap(batch) <= scheduler.cfg.MaxBatchPerWorker*2 {
+					batch = batch[:0]
+				} else {
+					batch = nil
 				}
-				// queueIdx is calculated from the last job's worker key in the batch
-				queueIdx := int(utils.GenHashKey(*msg.OutputStreamKey)) % nrWorkers
-
-				batchLen := len(realBatch)
-				theBatch := make([]*core.Msg, batchLen)
-				copy(theBatch, realBatch)
-
-				scheduler.workerQueues[queueIdx] <- theBatch
-				batch = nil
-				realBatch = nil
 			}
-
 		}
-	}(scheduler.buffers[tableKey], tableKey)
 
-	scheduler.buffers[tableKey] <- msg
+		var once sync.Once
 
-	return nil
+		for {
+			select {
+			case msg, ok := <-c:
+				if !ok {
+					once.Do(func() {
+						if len(batch) > 0 {
+							flushFunc()
+						}
+						closing = true
+						scheduler.bufferWg.Done()
+					})
+					continue
+				}
+				batch = append(batch, msg)
+				batchLen := len(batch)
+				qlen := len(c)
+				metrics.QueueLength.WithLabelValues(core.PipelineName, "table-buffer", key).Set(float64(qlen))
+				if scheduler.needFlush(qlen, batchLen, scheduler.cfg.MaxBatchPerWorker) {
+					flushFunc()
+				}
+
+			case h, ok := <-latchC:
+				if !ok {
+					scheduler.latchWg.Done()
+					return
+				}
+				latch[h]--
+				if latch[h] == 0 {
+					delete(latch, h)
+					if len(batch) > 0 && len(latchC) == 0 {
```

Comment:
Why do we need make sure `len(latchC) == 0`?

Reply:
it's just an optimization. if there're still locks hold, the next batch is very likely to fail. And for each new coming msg, the batch will be retried.